src/compiler/checker.ts

```diff
@@ -10305,11 +10400,17 @@ namespace ts {
 
         function getTypeAtPosition(signature: Signature, pos: number): Type {
             return signature.hasRestParameter ?
-                pos < signature.parameters.length - 1 ? getTypeOfSymbol(signature.parameters[pos]) : getRestTypeOfSignature(signature) :
-                pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length - 1) :
+                getParameterTypeAtIndex(signature, pos, signature.parameters.length, anyType);
         }
 
         function assignContextualParameterTypes(signature: Signature, context: Signature, mapper: TypeMapper) {
+            if (context.thisType) {
+                if (signature.declaration.kind !== SyntaxKind.ArrowFunction) {
+                    // do not contextually type thisType for ArrowFunction. 
+                    signature.thisType = context.thisType;
+                }
+            }

```

Comment:
There's a test for it on line 164 and following:
https://github.com/Microsoft/TypeScript/pull/6739/files#diff-f0430cf5aca89cd84e8c29f53dbecbb4R164

Briefly,

```ts
interface I { a: number; m(): number }
let impl: I = { a: 12, m() { return this.a } }
```